critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -0,0 +1,212 @@
+        InventoryUtil.assertNotNull(FileSyncJob.class, "resourceID", resourceID);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "storageAdapter", storageAdapter);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "artifactDAO", artifactDAO);
+
+        this.artifactID = artifactID;
+        this.resourceID = resourceID;
+        this.storageAdapter = storageAdapter;
+        this.artifactDAO = artifactDAO;
+    }
+
+    @Override
+    public void run() {
+        RegistryClient regClient = new RegistryClient();
+        log.debug("resource id: " + this.resourceID);
+        URL certURL = regClient.getServiceURL(this.resourceID, Standards.VOSPACE_SYNC_21, AuthMethod.CERT);
+        log.debug("certURL: " + certURL);
+
+        try {
+            // Ask for all protocols available back, and the server will
+            // give you URLs for the ones it allows.
+            List<Protocol> protocolList = new ArrayList<Protocol>();
+            Protocol httpsCert = new Protocol(VOS.PROTOCOL_HTTPS_GET);
+
+            log.debug("protocols: " + httpsCert);
+            protocolList.add(httpsCert);
+
+            Transfer transfer = new Transfer(artifactID, Direction.pullFromVoSpace, protocolList);
+            transfer.version = VOS.VOSPACE_21;
+
+            TransferWriter writer = new TransferWriter();
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            writer.write(transfer, out);
+            FileContent content = new FileContent(out.toByteArray(), "text/xml");
+            log.debug("xml file content to be posted: " + transfer);
+
+            log.debug("artifact path: " + artifactID.getPath());
+            HttpPost post = new HttpPost(certURL, content, true);
+            post.prepare();
+            log.debug("post prepare done");
+
+            // TODO: error handling is likely very different than this...
+            if (post.getThrowable() != null && post.getThrowable() instanceof FileNotFoundException) {
+                throw (FileNotFoundException) post.getThrowable();
+            }
+
+            TransferReader reader = new TransferReader();
+            Transfer t = reader.read(post.getInputStream(), null);
+            List<String> urlList = t.getAllEndpoints();
+
+            log.debug("endpoints returned: " + urlList);
+            // get the URL list from t, then call HttpGet to get the item
+
+            int retryCount = 0;
+            boolean found = false;
+
+            for (int i = 0; i < urlList.size(); i++) {
+                URL u = new URL(urlList.get(i));
+                log.debug("trying this url: " + u);
+
+                while (retryCount < 2 && found == false) {
+                    ByteArrayOutputStream dest = new ByteArrayOutputStream();
+                    HttpGet get = new HttpGet(u, dest);
+                    get.prepare();
+                    Artifact curArtifact = artifactDAO.get(artifactID);
+
```

Comment:
The above artifactDAO.get is too early (before the data transfer); it should happen farther down: right before curArtifact is used. That's not technically sufficient to prevent all race conditions, but it will make it easier to add the right locking later on.

Reply:
This has been moved, in the rework done in response to your previous comment.